website/addons/dryad/views.py

```diff
@@ -0,0 +1,225 @@
+from pyDryad import Dryad_DataOne, Dryad_SOLR
+
+import logging
+
+from flask import request
+from framework.flask import redirect
+from urllib2 import HTTPError
+
+from website.project.decorators import must_have_addon, must_be_valid_project
+from website.project.views.node import _view_project
+
+from website.util import rubeus
+
+from website.project.decorators import must_be_contributor_or_public
+
+logger = logging.getLogger(__name__)
+
+import xml.etree.ElementTree as ET
+
+@must_be_valid_project
+@must_have_addon('dryad', 'node')
+def check_dryad_doi(node_addon, **kwargs):
+    doi = request.args["doi"]
+    try:
+        d = Dryad_DataOne()
+        d.metadata(doi)
+    except HTTPError:
+        return {'result': False}
+    return {'result': True}
+
+
+@must_be_valid_project
+@must_have_addon('dryad', 'node')
+def set_dryad_doi(node_addon, **kwargs):
+    doi = request.args["doi"]
+    pid = kwargs['pid']
+    auth = kwargs['auth']
+    d = Dryad_DataOne()
+    try:
+        m = d.metadata(doi)
+    except HTTPError:
+        return redirect("/project/{}".format(pid))
+
+    node_addon.set_doi(doi, m.getElementsByTagName("dcterms:title")[0].firstChild.wholeText, auth)
+    node_addon.save()
+    #now, redirect back to the original homepage
+    return redirect("/project/{}".format(pid))
+
+@must_be_valid_project
+@must_have_addon('dryad', 'node')
+def remove_dryad_doi(node_addon, **kwargs):
+    doi = request.args["doi"]
+    node_addon.dryad_doi_list.remove(doi)
+
+    node_addon.save()
+    return {}
+
+@must_be_valid_project
+@must_have_addon('dryad', 'node')
+def dryad_browser(**kwargs):
+    return dryad_page(**kwargs)
+
+@must_be_valid_project
+@must_have_addon('dryad', 'node')
+def dryad_page(**kwargs):
+    node = kwargs['node']
+    pid = kwargs['pid']
+    auth = kwargs['auth']
+
+    count = 10
+    if "count" in request.args:
+        count = int(request.args["count"])
+    start = 0
+    if "start" in request.args:
+        start = int(request.args["start"])
+
+    dryad = node.get_addon('dryad')
+
+    d = Dryad_DataOne()
+    logger.info("Getting Package List from Dryad DataOne API")
+    x = d.list(count=count, start_n=start)
+    count = int(x.getElementsByTagName("d1:objectList")[0].attributes["count"].value)
+    start = int(x.getElementsByTagName("d1:objectList")[0].attributes["start"].value)
+    total = int(x.getElementsByTagName("d1:objectList")[0].attributes["total"].value)
+
+    ret = {"end": start + count,
+            "start": start,
+            "total": total,
+            "content": "",
+            'next_dryad': request.path + "?count={}&start={}".format(count, start + count),
+            'previous_dryad': request.path + "?count={}&start={}".format(count, start - count)}
+
+    objectList = ET.Element("ul")
+
+    for obj in x.getElementsByTagName("objectInfo"):
```

Comment:
It might be nice to move this code into a separate utility method for improved testability and readability.